archinfo/arch.py

```diff
@@ -199,37 +199,19 @@ def __init__(self, endness, instruction_endness=None):
                                 break
 
             self.default_register_values = [(r.name,) + r.default_value for r in self.register_list if r.default_value is not None]
-#           print 'default_register_values'
-#           for i in self.default_register_values:
-#               print i
             self.entry_register_values = {r.name: r.linux_entry_value for r in self.register_list if r.linux_entry_value is not None}
-#           print 'entry_register_values'
-#           for i in self.entry_register_values:
-#               print i
             self.default_symbolic_registers = [r.name for r in self.register_list if r.general_purpose]
-#           print 'default_symbolic_registers'
-#           for i in self.default_symbolic_registers:
-#               print i
             self.register_names = {r.vex_offset: r.name for r in self.register_list}
-#           print 'register_names'
-#           for k, v in self.register_names.iteritems():
-#               print k, v
             self.registers = self._get_register_dict()
-#           print 'registers'
-#           for k, v in self.registers.iteritems():
-#               print k, v
             self.argument_registers = set(r.vex_offset for r in self.register_list if r.argument)
-#           print 'argument_registers'
-#           for i in self.argument_registers:
-#               print i
             self.persistent_regs = [r.name for r in self.register_list if r.persistent]
-#           print 'persistent_regs'
-#           for i in self.persistent_regs:
-#               print i
             self.concretize_unique_registers = set(r.vex_offset for r in self.register_list if r.concretize_unique)
-#           print 'concretize_unique_registers'
-#           for i in self.concretize_unique_registers:
-#               print i
+
+            # Register offsets
+            self.ip_offset = self.registers['ip'][0]
+            self.sp_offset = self.registers['sp'][0]
+            self.bp_offset = self.registers['bp'][0]
+            self.lr_offset = self.registers.get('lr', (None, None))[0]
```

Comment:
I would do this by adding attributes to the Register class - `instruction_pointer=True, stack_pointer=True, base_pointer=True, link_register=True`. Relying on these names seems sketchy to me.

Reply:
However, there are a lot of other sketchy things going on in archinfo, and this seems like a pretty safe change, so I'll merge it for now.

Reply:
What are other sketchy things? I can try to fix them! But I think since these names are used by other modules (`angr`, `cle`, `pyvex`, ...), they gotta be safe enough?

Reply:
Yeah, I would prefer angr, cle etc didn't use `state.regs.sp` etc in order to get the stack pointer. Ideally `state.sp` would do `state.registers.load(state.arch.sp_offset, state.arch.bits)`. maybe archinfo should provide the size so we don't have to guess with state.arch.bits either?

All this stuff is sketchy - you run into it immediately when you try to implement some exotic architecture like AVR that has different sizes for different pointers, multiple memories, etc. I definitely can't just give you a list...